xCAT-client-2.0/share/man/man3/noderange.3
.\" Process this file with
.\" groff -man -Tascii noderange.3
.\"
.TH NODERANGE 3 "JUNE 2002" xCAT "xCAT"
.SH NAME
noderange \- generate a list of node names
.SH SYNOPSIS
.B noderange
[\fIcomma delimited nodelists and grouplists\fR]
.br

.I e.g.
.br

node1,node2,node8,node20,group1
.br

node14-node56,node70-node203,group1-group10
.br

node1,node2,node8,node20,node14-node56,node70-node203
.br

node[14-56]
.br

all,-node129-node256,-frame01-frame03
.br

/node.*
.br

^/tmp/nodes
.br

node10+5
.br

10-15,-13
.br

group1@group2
.br

.br
.SH DESCRIPTION
.B noderange
is a syntax that can be used in most xCAT commands to conveniently
specify a list of nodes.  The result is that the command will
be applied to a range of nodes, often in parallel.

.B noderange
is a comma-separate list.  Each token (text between commas) in the list can be any
of the forms listed below:

Individual node or group:

.I node01
.br
.I group1
.br

A range of nodes or groups:

.I node01-node10
(equivalent to: node01,node02,node03,...node10)
.br
.I group1-group3
(equivalent to: group1,group2,group3)
.br

A regular expression match of nodes or groups:

.I /node[345].*
(will match any nodes that start with node3, node4, or node5)
.br
.I /group[12].*
(will match any groups that start with group1 or group2)
.br

An incremented range of nodes:

.I node10+3
(equivalent to: node10,node11,node12,node13)
.br

The full path of a file containing noderanges of nodes or groups:

.I ^/tmp/nodelist
.br

A node shorthand range of nodes:

.I 10-20
(if prefix is 'node', equivalent to: node10,node11,node12,...node20)
.br
.I 10+3
(if prefix is 'node', equivalent to: node10,node11,node12,node13)
.br

Or any combination:

.I node01-node30,node40,^/tmp/nodes,/node[13].*,2-10,node50+5
.br

Any individual
.B noderange
may be prefixed with an exclusion operator
(default -) with the exception of the file operator (default ^).

The intersection operator @ calculates the intersection of the left and right sides:

.I group1@group2
(will result in the list of nodes group1 and group2 have in common)

Any combination or multiple combinations of inclusive
and exclusive ranges of nodes and groups is legal.  There
is no precedence implied in the order of the
arguments.  Exclusive ranges have precedence over inclusive.

Nodes have precedence over groups.  If a node range match is made
then no group range match will be attempted.

All node names are validated against
the nodelist table.
Invalid nodes are ignored and return nothing.

All group names are validated against
the nodelist and nodetype table.
Invalid groups are ignored and return nothing.

Throughout this
man page the term
\fIxCAT Node Format\fR
will be used.

\fIxCAT Node Format\fR is defined by the following regex:

^([A-Za-z-]+)([0-9]+)(([A-Za-z-]+[A-Za-z0-9-]*)*)

In plain english a node or group is in
\fIxCAT Node Format\fR
if starting from the begining there are one or more
alpha characters of any case and any number of - in any
combination, followed by
one or more numbers, then optionally followed by one
alpha character of any case or the - followed by any
combination of case mixed alphanumerics and the -.

.B noderange
supports node/group names in
.I any
format.
\fIxCAT Node Format\fR
is
.B not
required, however some node range
methods used to determine range will not be used.
E.g. If using a
.B noderange
of
.I node1a-node9a
with a
.B nodelist
table only listing
.I node1a
through
.IR node5a ,
.B noderange
will enumerate then validate and return a proper
range.  If using a node range of
.I aa-az
with a
.B nodelist
table only listing
.I aa
through
.IR ay,
.B noderange
will fail to return any values.

Example of
\fIxCAT Node Format\fR
node/group names:

.B nodename			prefix	number	suffix
.br
node1			node		1
.br
node001			node		001
.br
node-001			node-	001
.br
node-foo-001-bar	node-foo-	001		-bar
.br
node-foo-1bar		node-foo-	1		bar
.br
foo1bar2			foo		1		bar2
.br
rack01unit34		rack		01		unit34
.br
unit34rack01		unit		34		rack01
.br
pos0134			pos		0134
.br

Example of non-\fIxCAT Node Format\fR
node/group names, but still valid:

aa
.br
yellow
.br
red
.br
12foo
.br


The supported
.B noderange
syntaxes are checked for, and processed, in a specific order.  First
.B noderange
checks for the multiple range operator (default ,).  Each range is
also processed by
.BR noderange .

Next
.B noderange
checks for the file operator (default ^).  If the file exists (must be a full path name)
each line will be processed as a
.BR noderange .
Lines starting
with
.I #
or the file operator (default ^) are ignored.  Only
one
.B noderange
per line is read.  All characters are ignored
after the first white space.

e.g.

^/tmp/nodes

where

cat /tmp/nodes outputs:

#my node list (this line ignored)
.br
^/tmp/foo #ignored
.br
node01    #node comment
.br
node02
.br
node03
.br
node10-node20
.br
/group[456].*
.br
-node50
.br

Next
.B noderange
checks for the exclusion operator (default -) then continues.
This operator supports nodes and groups.
.B noderange
will not confuse the exclusion or range
operators with the
- character in names.

Next
.B noderange
checks for a numeric only range (e.g. 10-20, 5+3, or just 10), then uses
.I $XCAT_NODE_PREFIX
(default is 'node') and
.I $XCAT_NODE_SUFFIX
(optional)
as the defaults to complete the node names.
.B nodeRange
tries to be intelligent about detecting padding, so you can specify 'node001-node200'
and it will add the proper number of zeroes to make all numbers 3 digits.
Noderange shorthand supports nodes only.  Noderange shorthand can
be mixed with all other operators except regex.  i.e.  exclusion,
increment, range, and file may be used.

Next
.B noderange
checks for the for the regular expression operator (default /).
Regular expressions offer the most flexibility.  If you are
interested in learning regex read the book
\fIMastering Regular Expressions\fR.
This operator supports nodes and groups.

Next
.B noderange
checks for the increment range operator (default +).  Increment
range operator noderanges are in the format:

.IB valid_node_name + number_of_sequential_nodes

e.g.

.I node10+5

would yield node10 plus the next
.I 5
nodes.

This action is performed using two different methods.
If
.I valid_node_name
is in
\fIxCAT Node Format\fR
then the range is enumerated to
.IR number_of_sequential_nodes .
If not in
\fIxCAT Node Format\fR
then a sorted
.B nodelist
table is used to return the node range.
This operator supports nodes only.

Next
.B noderange
checks for a single node name or group name.

Next
.B noderange
checks for the range operator (default -).  Ranges are
performed first by validating that both the start
and end nodes or groups defining the range exist and if
so the range is returned based on the content of the
.B nodelist
and
.B nodetype
tables.
If the start and end nodes or groups
defined in the range do not exist,
.I and
if both are in
\fIxCAT Node Format\fR,
.I and
if both the prefix and suffix match,
then the range is enumerated and each node/group
validated.  Only valid nodes/groups will be returned.
.B noderange
will not confuse the exclusion or range
operators with the
- character in names.

Last
.B noderange
returns nothing if no match can be found.

.B noderange
uses the smallest integer to determine
padding.  e.g.  \fInode1-node10\fR will generate a list of nodes
with numbers 1, 2, 3, 4, 5, 6, 7, 8, 9, 10.  \fInode001-node010\fR will
generate a list of nodes with numbers 001, 002, 003, 004,
005, 006, 007, 008, 009, 010.

Multiple instances of a node name are treated as one
instance.  e.g. \fInode1-node10,node4,node4,node4\fR will generate
a list of nodes numbered 1 through 10, the number 4
will only be listed once.

.SH EXAMPLES
.I all,-node5-node10

Generates a list of all nodes (assuming all is a group)
listed in the
.B nodelist
table less node5 through node10.

.I node1-node10,-node3-node5,node4

Generates a list of nodes 1 through 10 less nodes 3,4,5.  Note
that node4 is listed twice, first in the range and then at
the end.  Because exclusion has precedence node4 will be excluded.

.I node1-node10,-node3,-node5

Generates a list of nodes 1 through 10 less nodes 3 and 5.

.I -node17-node32,all

Generates a list of all (assuming 'all' is a group) nodes in the
.B nodelist
table less 17 through 32.

.I node1-node128,user1-user4

Generates a list of nodes 1 through 128, and user nodes 1 through 4.

.I all,-rack1-rack3,-node100-node200,node150,-storage

Generates a list of all nodes (assuming 'all' is a group), less
nodes in groups rack1 through rack3 (assuming groups rack1,
rack2, and rack3 are defined), less nodes 100 through
200, less nodes in the storage group.  Note that
node150 is listed but is excluded.

.I /node[23].*

Generates a list of nodes matching the regex
.IR node[23].* .
That is all nodes that start with node2 or node3 and end in anything
or nothing.  E.g. node2, node3, node20, node30, node21234 all match.

.SH "NOTES"
.IX Header "NOTES"
This command is part of the xCAT software.


perl-xCAT-2.0/xCAT/NodeRange.pm
# IBM(c) 2007 EPL license http://www.eclipse.org/legal/epl-v10.html
package xCAT::NodeRange;
use xCAT::Table;
use Data::Dumper;
require Exporter;
use strict;

#Perl implementation of noderange
our @ISA = qw(Exporter);
our @EXPORT = qw(noderange nodesmissed);

my $missingnodes=[];
my $nodelist; #=xCAT::Table->new('nodelist',-create =>1);
#my $nodeprefix = "node";


sub subnodes (\@@) {
    #Subtract set of nodes from the first list
    my $nodes = shift;
    my $node;
    foreach $node (@_) {
        @$nodes = (grep(!/^$node$/,@$nodes));
    }
}
sub nodesmissed {
  return @$missingnodes;
}

sub expandatom {
	my $atom = shift;
	my $verify = (scalar(@_) == 1 ? shift : 1);
        my @nodes= ();
    #TODO: these env vars need to get passed by the client to xcatd
	my $nprefix=(defined ($ENV{'XCAT_NODE_PREFIX'}) ? $ENV{'XCAT_NODE_PREFIX'} : 'node');
	my $nsuffix=(defined ($ENV{'XCAT_NODE_SUFFIX'}) ? $ENV{'XCAT_NODE_SUFFIX'} : '');
	if ($nodelist->getAttribs({node=>$atom},'node')) {		#The atom is a plain old nodename
		return ($atom);
	}
    if ($atom =~ /^\(.*\)$/) {     # handle parentheses by recursively calling noderange()
      $atom =~ s/^\((.*)\)$/$1/;
      return noderange($atom);
    }

    # Try to match groups?
	foreach($nodelist->getAllAttribsWhere("groups like '%".$atom."%'",'node','groups')) {
		my @groups=split(/,/,$_->{groups}); #The where clause doesn't guarantee the atom is a full group name, only that it could be
		if (grep { $_ eq "$atom" } @groups ) {
			push @nodes,$_->{node};
		}
	}

	if ($atom =~ m/^[0-9]+\z/) {    # if only numbers, then add the prefix
		my $nodename=$nprefix.$atom.$nsuffix;
		return expandatom($nodename,$verify);
	}
	my $nodelen=@nodes;
	if ($nodelen > 0) {
		return @nodes;
	}

	if ($atom =~ m/^\//) { # A regular expression
        unless ($verify) { # If not in verify mode, regex makes zero possible sense
          return ($atom);
        }
		#TODO: check against all groups
		$atom = substr($atom,1);
		foreach ($nodelist->getAllAttribs('node')) {
			if ($_->{node} =~ m/^${atom}$/) {
				push(@nodes,$_->{node});
			}
		}
		return(@nodes);
	}

	if ($atom =~ m/(.*)\[(.*)\](.*)/) { # square bracket range
	#for the time being, we are only going to consider one [] per atom
	#xcat 1.2 does no better
		my @subelems = split(/([\,\-\:])/,$2);
		my $subrange="";
		while (my $subelem = shift @subelems) {
			my $subop=shift @subelems;
			$subrange=$subrange."$1$subelem$3$subop";
		}
		foreach (split /,/,$subrange) {
			my @newnodes=expandatom($_,$verify);
			@nodes=(@nodes,@newnodes);
		}
		return @nodes;
	}

	if ($atom =~ m/\+/) {  # process the + operator
		$atom =~ m/^([^0-9]*)([0-9]+)([^\+]*)\+([0-9]+)/;
		my $pref=$1;
		my $startnum=$2;
		my $suf=$3;
		my $end=$4+$startnum;
        my $endnum = sprintf("%d",$end);
        if (length ($startnum) > length ($endnum)) {
          $endnum = sprintf("%0".length($startnum)."d",$end);
        }
		if (($pref eq "") && ($suf eq "")) {
			$pref=$nprefix;
			$suf=$nsuffix;
		}
		foreach ("$startnum".."$endnum") {
			my @addnodes=expandatom($pref.$_.$suf,$verify);
			@nodes=(@nodes,@addnodes);
		}
		return (@nodes);
	}

    if ($atom =~ m/[-:]/) { # process the minus range operator
      my $left;
      my $right;
      if ($atom =~ m/:/) {
        ($left,$right)=split /:/,$atom;
      } else {
        my $count= ($atom =~ tr/-//);
        if (($count % 2)==0) { #can't understand even numbers of - in range context
          if ($verify) {
            push @$missingnodes,$atom;
            return ();
          } else { #but we might not really be in range context, if noverify
            return  ($atom);
          }
        }
        my $expr="([^-]+?".("-[^-]*"x($count/2)).")-(.*)";
        $atom =~ m/$expr/;
        $left=$1;
        $right=$2;
      }
      if ($left eq $right) { #if they said node1-node1 for some strange reason
		return expandatom($left,$verify);
      }
      my @leftarr=split(/(\d+)/,$left);
      my @rightarr=split(/(\d+)/,$right);
      if (scalar(@leftarr) != scalar(@rightarr)) { #Mismatch formatting..
        if ($verify) {
          push @$missingnodes,$atom;
          return (); #mismatched range, bail.
        } else { #Not in verify mode, just have to guess it's meant to be a nodename
          return  ($atom);
        }
      }
      my $prefix = "";
      my $suffix = "";
      foreach (0..$#leftarr) {
        my $idx = $_;
        if ($leftarr[$idx] =~ /^\d+$/ and $rightarr[$idx] =~ /^\d+$/) { #pure numeric component
          if ($leftarr[$idx] ne $rightarr[$idx]) { #We have found the iterator (only supporting one for now)
            my $prefix = join('',@leftarr[0..($idx-1)]); #Make a prefix of the pre-validated parts
            my $luffix; #However, the remainder must still be validated to be the same
            my $ruffix;
            if ($idx eq $#leftarr) {
              $luffix="";
              $ruffix="";
            } else {
              $ruffix = join('',@rightarr[($idx+1)..$#rightarr]);
              $luffix = join('',@leftarr[($idx+1)..$#leftarr]);
            }
            if ($luffix ne $ruffix) { #the suffixes mismatched..
              if ($verify) {
                push @$missingnodes,$atom;
                return ();
              } else {
                return ($atom);
              }
            }
            foreach ($leftarr[$idx]..$rightarr[$idx]) {
              my @addnodes=expandatom($prefix.$_.$luffix,$verify);
              @nodes=(@nodes,@addnodes);
            }
            return (@nodes); #the return has been built, return, exiting loop and all
          }
        } elsif ($leftarr[$idx] ne $rightarr[$idx]) {
          if ($verify) {
            push @$missingnodes,$atom;
            return ();
          } else {
            return ($atom);
          }
        }
        $prefix .= $leftarr[$idx]; #If here, it means that the pieces were the same, but more to come
      }
      #I cannot conceive how the code could possibly be here, but whatever it is, it must be questionable
      if ($verify) {
        push @$missingnodes,$atom;
        return (); #mismatched range, bail.
      } else { #Not in verify mode, just have to guess it's meant to be a nodename
        return  ($atom);
      }
	}

    push @$missingnodes,$atom;
	if ($verify) {
		return ();
	} else {
		return ($atom);
	}
}

sub noderange {
  $missingnodes=[];
  #We for now just do left to right operations
  #TODO: Parentheses... A parenthetical group to the right of an intersection makes the obvious
  #answer not work
  my $range=shift;
  my $verify = (scalar(@_) == 1 ? shift : 1);
  $nodelist =xCAT::Table->new('nodelist',-create =>1);
  my %nodes = ();
  my %delnodes = ();
  my $op = ",";
  #my @elems = split(/(,(?![^[]*?])|@)/,$range); #, or @ but ignore , within []
  my @elems = split(/(,(?![^[]*?])(?![^\(]*?\))|@(?![^\(]*?\)))/,$range);  # comma or @ but ignore comma within []

  while (my $atom = shift @elems) {
    if ($atom =~ /^-/) {           # if this is an exclusion, strip off the minus, but remember it
      $atom = substr($atom,1);
      $op = $op."-";
    }

    if ($atom =~ /^\^(.*)$/) {    # get a list of nodes from a file
      open(NRF,$1);
      while (<NRF>) {
        my $line=$_;
        unless ($line =~ m/^[\^#]/) {
          $line =~ m/^([^:	 ]*)/;
          my $newrange = $1;
          chomp($newrange);
          my @filenodes = noderange($newrange);
          foreach (@filenodes) {
            $nodes{$_}=1;
          }
        }
      }
      close(NRF);
      next;
    }

    my %newset = map { $_ =>1 } expandatom($atom,$verify);    # expand the atom and make each entry in the resulting array a key in newset

    if ($op =~ /@/) {       # compute the intersection of the current atom and the node list we have received before this
      foreach (keys %nodes) {
        unless ($newset{$_}) {
          delete $nodes{$_};
        }
      }
    } elsif ($op =~ /,-/) {        # add the nodes from this atom to the exclude list
		foreach (keys %newset) {
			$delnodes{$_}=1; #delay removal to end
		}
	} else {          # add the nodes from this atom to the total node list
		foreach (keys %newset) {
			$nodes{$_}=1;
		}
	}
	$op = shift @elems;

    }    # end of main while loop

    # Now remove all the exclusion nodes
    foreach (keys %nodes) {
		if ($delnodes{$_}) {
			delete $nodes{$_};
		}
    }
    undef $nodelist;
    return sort (keys %nodes);

}


1;

=head1 NAME

xCAT::NodeRange - Perl module for xCAT noderange expansion

=head1 SYNOPSIS

	use xCAT::NodeRange;
	my @nodes=noderange("storage@rack1,node[1-200],^/tmp/nodelist,node300-node400,node401+10,500-550");

=head1 DESCRIPTION

noderange interprets xCAT noderange formatted strings and returns a list of xCAT nodelists.  The following two operations are supported on elements, and interpreted left to right:

, union next element with everything to the left.

@ take intersection of element to the right with everything on the left (i.e. mask out anything to the left not belonging to what is described to the right)

Each element can be a number of things:

A node name, i.e.:

=item * node1

A hyphenated node range (only one group of numbers may differ between the left and right hand side, and those numbers will increment in a base 10 fashion):

node1-node200 node1-compute-node200-compute
node1:node200 node1-compute:node200-compute

A noderange denoted by brackets:

node[1-200] node[001-200]

A regular expression describing the noderange:

/d(1.?.?|200)

A node plus offset (this increments the first number found in nodename):

node1+199

And most of the above substituting groupnames.
3C
3C

NodeRange tries to be intelligent about detecting padding, so you can:
node001-node200
And it will increment according to the pattern.


=head1 AUTHOR

Jarrod Johnson (jbjohnso@us.ibm.com)

=head1 COPYRIGHT

Copyright 2007 IBM Corp.  All rights reserved.


=cut
